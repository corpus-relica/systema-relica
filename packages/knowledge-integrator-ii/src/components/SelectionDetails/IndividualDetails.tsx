const IndividualDetails = () => {
  return (
    <div>
      <h1>Individual Details</h1>
    </div>
  );
};

export default IndividualDetails;
