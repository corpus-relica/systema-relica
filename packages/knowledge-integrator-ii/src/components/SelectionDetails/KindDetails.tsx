import React, { useState, useEffect } from "react";
import { useStore } from "react-admin";
import { useQuery } from "@tanstack/react-query";
import { retrieveKindModel } from "../../CCClient";
import { sockSendCC } from "../../socket";
import Box from "@mui/material/Box";
import IconButton from "@mui/material/IconButton";
import CopyAllIcon from "@mui/icons-material/CopyAll";
import Stack from "@mui/material/Stack";
import Typography from "@mui/material/Typography";

import Specialization from "./display/Specialization";
import Definition from "./display/Definition";

const KindDetails: React.FC = () => {
  const [selectedNode] = useStore("selectedNode");

  const { isLoading, error, data } = useQuery({
    queryKey: ["kindModel", selectedNode],
    queryFn: () =>
      selectedNode
        ? retrieveKindModel(selectedNode).then((res) => res)
        : Promise.resolve(null),
    enabled: !!selectedNode, // This disables the query if selectedNode is null
  });

  if (!selectedNode) return <div>No node selected</div>;
  if (isLoading) return <div>Loading...</div>;
  if (error) return <div>Error: {error.message}</div>;

  const { uid, type, category, name, definition, facts, collection } = data;
  const specialization = data[1146];
  const synonyms = data[1981];
  const inverses = data[1986];
  const reqRole1 = data[4731];
  const reqRole2 = data[4733];
  const possRoles = data[4714];

  const loadAllPossRoles = (uid) => {
    possRoles.forEach((possRole) => {
      sockSendCC("user", "loadEntity", { uid: possRole });
    });
  };

  const pushDataToClipboard = async () => {
    await window.navigator.clipboard.writeText(JSON.stringify(data));
  };

  const text = `${uid} (${type}) :: ${name}`;
  // const factTableRows = facts.map((fact) => {
  //   return (
  //     <TableRow>
  //       <TableCell>
  //         <Text size="10px">{fact.lh_object_name}</Text>
  //       </TableCell>
  //       <TableCell>
  //         <Text size="10px">{fact.rel_type_name}</Text>
  //       </TableCell>
  //       <TableCell>
  //         <Text size="10px">{fact.rh_object_name}</Text>
  //       </TableCell>
  //     </TableRow>
  //   );
  // });

  console.log("data: ", data);

  return (
    <Stack direction="column" spacing="1">
      <Box>
        <Stack direction="row" spacing="1">
          <Box>
            <Typography size="18px" style={{ fontWeight: 800, color: "black" }}>
              {text}
            </Typography>
            <Typography size="12px" style={{ fontWeight: 500, color: "black" }}>
              Specialization:
            </Typography>
            {specialization && specialization.length > 0 && (
              <Specialization uids={specialization} childUID={uid} />
            )}
          </Box>
          <IconButton aria-label="copy to clipboard">
            <CopyAllIcon onClick={pushDataToClipboard} />
          </IconButton>
        </Stack>
        {category && (
          <Stack direction="row" spacing="1">
            <Typography size="12px" style={{ fontWeight: 600 }}>
              Category:
            </Typography>
            <Typography size="12px">{category}</Typography>
          </Stack>
        )}
        {definition && <Definition definitions={definition} />}
      </Box>
    </Stack>
  );
};

export default KindDetails;

// <Box
//   sx={{
//     flexDirection: "column",
//     width: "384px",
//   }}
// >
//   <Box>
//     <Box sx={{ flexDirection: "row" }}>
//       <Box></Box>
//       <Button onClick={pushDataToClipboard} variant="outlined">
//         push to clipboard
//       </Button>
//     </Box>
//   </Box>
// </Box>
