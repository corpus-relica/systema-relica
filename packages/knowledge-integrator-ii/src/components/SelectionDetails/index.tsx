import React, { useState, useEffect } from "react";
import { useStore } from "react-admin";

const SelectionDetails = () => {
  const [selectedNode] = useStore("selectedNode");
  const [selectedEdge] = useStore("selectedEdge");

  useEffect(() => {
    // if (selectedNode) {
    //   setNode(selectedNode);
    // }
    console.log("FOO", selectedNode);
  }, [selectedNode]);

  return (
    <>
      <h1>Selection Details</h1>
      {selectedNode}
      {selectedEdge}
    </>
  );
};

export default SelectionDetails;
