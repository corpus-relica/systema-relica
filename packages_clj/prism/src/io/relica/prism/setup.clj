(ns io.relica.prism.setup
  (:require [taoensso.timbre :as log]
            [io.relica.prism.db :as db]
            [io.relica.prism.xls :as xls]
            [io.relica.prism.config :as config]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [next.jdbc :as jdbc]
            [next.jdbc.result-set :as rs]
            [buddy.hashers :as hashers]))

;; Setup state tracking
(defonce setup-state (atom {:stage :not-started
                           :master-user nil
                           :status "Setup has not been initiated."
                           :progress 0
                           :error nil}))

;; Setup stages
(def setup-stages [:db-check :user-setup :db-seed :cache-build :complete])

(defn get-setup-state 
  "Returns the current setup state for external APIs."
  []
  @setup-state)

(defn advance-stage!
  "Advances the setup to the next stage."
  []
  (let [current-stage (:stage @setup-state)
        current-index (when (keyword? current-stage)
                       (.indexOf setup-stages current-stage))
        next-index (inc current-index)
        next-stage (if (and (number? next-index) (< next-index (count setup-stages)))
                     (nth setup-stages next-index)
                     :complete)]
    (swap! setup-state assoc
           :stage next-stage
           :progress (if (= next-stage :complete) 100
                       (* (/ (inc next-index) (count setup-stages)) 100)))))

(defn update-status!
  "Updates the setup status message."
  [status-message]
  (swap! setup-state assoc :status status-message)
  (log/info status-message)
  ;; We'll add a call to broadcast the update via WebSocket
  ;; This will be implemented later and injected to avoid circular dependencies
  (when-let [broadcast-fn (resolve 'io.relica.prism.websocket/broadcast-setup-update)]
    (broadcast-fn)))

(defn set-error!
  "Sets an error in the setup state."
  [error-message]
  (log/error error-message)
  (swap! setup-state assoc :error error-message))

(defn start-setup-sequence!
  "Initiates the interactive setup sequence."
  []
  (log/info "Starting setup sequence...")
  (swap! setup-state assoc 
         :stage :db-check 
         :error nil
         :status "Checking database state..."
         :progress 10)
  
  (try
    (if (db/database-empty?)
      (do
        (update-status! "Database is empty. Ready for initial setup.")
        (advance-stage!)) ; Move to :user-setup
      (do
        (update-status! "Database already contains data. Setup not required.")
        (swap! setup-state assoc :stage :complete :progress 100)))
    (catch Exception e
      (set-error! (str "Error checking database state: " (.getMessage e))))))

(defn validate-credentials
  "Validates the admin user credentials."
  [username password confirm-password]
  (cond
    (str/blank? username)
    {:valid false :message "Username cannot be blank"}
    
    (< (count username) 4)
    {:valid false :message "Username must be at least 4 characters"}
    
    (str/blank? password)
    {:valid false :message "Password cannot be blank"}
    
    (< (count password) 8)
    {:valid false :message "Password must be at least 8 characters"}
    
    (not= password confirm-password)
    {:valid false :message "Passwords do not match"}
    
    :else
    {:valid true}))

(defn create-admin-user!
  "Creates the admin user with the given credentials."
  [username password]
  (log/info "Creating admin user:" username)
  (update-status! (str "Creating admin user: " username))
  
  (try
    ;; Create PostgreSQL admin user
    (let [email (str username "@relica.io") ; Generate an email based on username
          password-hash (buddy.hashers/derive password {:algorithm :bcrypt})
          jdbc-conn (jdbc/get-connection (config/db-spec))
          
          ;; Create user
          user-result (jdbc/execute-one! jdbc-conn
                     ["INSERT INTO users
                       (email, username, password_hash, is_active, is_admin, first_name, last_name)
                       VALUES (?, ?, ?, true, true, 'Admin', 'User')
                       RETURNING id, email, username, is_active, is_admin"
                      email username password-hash]
                     {:builder-fn next.jdbc.result-set/as-unqualified-maps})
          
          ;; Create a default environment
          env-result (when user-result
                       (jdbc/execute-one! jdbc-conn
                        ["INSERT INTO environments (name) 
                          VALUES ('Default Environment')
                          RETURNING id"]
                        {:builder-fn next.jdbc.result-set/as-unqualified-maps}))
          
          ;; Link user to environment
          user-env-result (when (and user-result env-result)
                            (jdbc/execute-one! jdbc-conn
                              ["INSERT INTO user_environments
                                (user_id, environment_id, is_owner, can_write)
                                VALUES (?, ?, true, true)
                                RETURNING id"
                               (:id user-result) (:id env-result)]
                              {:builder-fn next.jdbc.result-set/as-unqualified-maps}))]
      
      (log/info "Created admin user in PostgreSQL:" (dissoc user-result :password_hash))
      (when env-result (log/info "Created default environment with ID:" (:id env-result)))
      (when user-env-result (log/info "Linked user to environment"))
      
      (if (and user-result env-result user-env-result)
        (do
          (update-status! (str "Admin user " username " created successfully with default environment"))
          (swap! setup-state assoc :master-user username)
          (advance-stage!) ; Move to :db-seed
          true)
        (do
          (set-error! (str "Failed to create admin user"
                          (when (and user-result (not env-result))
                            " - environment creation failed")
                          (when (and user-result env-result (not user-env-result))
                            " - user-environment link failed")))
          false)))
    (catch Exception e
      (set-error! (str "Error creating admin user: " (.getMessage e)))
      false)))

(defn seed-database!
  "Seeds the database with initial data from XLS files."
  []
  (log/info "Starting database seeding process...")
  (update-status! "Processing seed files...")
  
  (try
    (let [csv-file-paths (xls/process-seed-directory)]
      (if (seq csv-file-paths)
        (do
          (update-status! (str "Processed " (count csv-file-paths) " seed files. Importing data..."))
          
          (let [total-files (count csv-file-paths)
                progress-increment (/ 40 total-files)] ; 40% of progress bar for this stage
            
            (doseq [[idx csv-path] (map-indexed vector csv-file-paths)
                    :let [file-name (.getName (io/file csv-path))]]
              
              (update-status! (str "Importing " file-name " (" (inc idx) "/" total-files ")..."))
              
              ;; Update progress after each file
              (swap! setup-state update :progress + progress-increment)
              
              ;; Create nodes
              (update-status! (str "Creating entities from " file-name "..."))
              (let [node-result (db/load-nodes-from-csv! file-name)]
                (when-not (:success node-result)
                  (throw (ex-info (str "Failed to load nodes from " file-name) 
                                  {:file file-name :error (:error node-result)}))))
              
              ;; Create relationships
              (update-status! (str "Creating relationships from " file-name "..."))
              (let [rel-result (db/load-relationships-from-csv! file-name)]
                (when-not (:success rel-result)
                  (throw (ex-info (str "Failed to load relationships from " file-name)
                                  {:file file-name :error (:error rel-result)}))))))
          
          (update-status! "Database seeding completed successfully")
          (advance-stage!) ; Move to :cache-build
          true)
        (do
          (update-status! "No seed files found. Database seeding skipped.")
          (log/warn "XLS processing did not generate any CSV files. Setup cannot proceed to cache building.")
          (log/warn "Setup will remain in the db-seed stage until seed files are processed correctly.")
          ;; Do NOT advance stage when no seed files are found - this is an error condition
          false)))
    (catch Exception e
      (set-error! (str "Error during database seeding: " (.getMessage e)))
      false)))

(defn build-caches!
  "Builds necessary caches for the application."
  []
  (log/info "Building database caches...")
  (update-status! "Building database caches...")
  
  (try
    ;; Execute cache building queries
    (let [subtype-cache-result (db/execute-query! 
                              "MATCH (sub:Entity)-[:role]->(:Fact {rel_type_uid: 1146})-[:role]->(super:Entity)
                               WITH sub, super
                               MERGE (sub)-[:SUBTYPE_OF]->(super)
                               RETURN count(*) as cache_count"
                              {})
          
          ;; Add more cache building queries here as needed
          ]
      
      (if (:success subtype-cache-result)
        (do
          (update-status! "Database caches built successfully")
          (advance-stage!) ; Move to :complete
          true)
        (do
          (set-error! "Failed to build database caches")
          false)))
    (catch Exception e
      (set-error! (str "Error building database caches: " (.getMessage e)))
      false)))

(defn complete-setup!
  "Marks the setup as complete."
  []
  (log/info "Setup completed successfully.")
  (update-status! "Setup completed successfully.")
  (swap! setup-state assoc :stage :complete :progress 100))

(defn handle-setup-stage!
  "Handles the current setup stage based on the state."
  []
  (let [current-stage (:stage @setup-state)]
    (case current-stage
      :not-started (start-setup-sequence!)
      :db-check (start-setup-sequence!) ; Re-run the check if needed
      :user-setup nil ; Wait for user input via API
      :db-seed (seed-database!)
      :cache-build (build-caches!)
      :complete (complete-setup!)
      ;; Default case
      (log/warn "Unknown setup stage:" current-stage))))

;; Initialize the setup
(defn init!
  "Initializes the setup module."
  []
  (log/info "Setup module initialized"))
