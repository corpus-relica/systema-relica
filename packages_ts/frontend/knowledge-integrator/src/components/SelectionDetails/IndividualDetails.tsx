import React, { useContext, useState, useEffect } from "react";
import { useStore } from "react-admin";
import { useQuery } from "@tanstack/react-query";

import Grid from "@mui/material/Grid";
import Box from "@mui/material/Box";
import Stack from "@mui/material/Stack";
import Typography from "@mui/material/Typography";
import IconButton from "@mui/material/IconButton";
import CopyAllIcon from "@mui/icons-material/FileCopy";

import Collection from "./display/Collection";
import IndividualName from "./display/IndividualName";
import Classification from "./display/Classification";
import Value from "./display/Value";
import Definition from "./display/Definition";
import WorkflowFactsVisualizer from "../../pages/Workflows/WorkflowFactsVisualizer";
import Synonyms from "./display/Synonyms";

import { portalClient } from "../../io/PortalClient.js";

const IndividualDetails: React.FC = () => {
  const [selectedNode] = useStore("selectedNode");

  const { isLoading, error, data } = useQuery({
    queryKey: ["individualModel", selectedNode],
    queryFn: () =>
      selectedNode
        ? portalClient.retrieveIndividualModel(selectedNode).then((res) => res)
        : Promise.resolve(null),
    enabled: !!selectedNode, // This disables the query if selectedNode is null
  });

  return <div>UNDER CONSTZRUCTION</div>
  if (isLoading || !selectedNode) return <div>Loading...</div>;
  if (error) return <div>Error: {error.message}</div>;

  const { uid, type, category, name, definition, facts, value, collection } =
    data;

  const classFact = facts.find((fact) => fact.rel_type_uid === 1225);
  const synFacts = facts.filter((fact) => fact.rel_type_uid === 1981);

  const specialization = data[1146];
  const classification = data[1225];
  const synonyms = data[1981];
  const inverses = data[1986];
  const reqRole1 = data[4731];
  const reqRole2 = data[4733];
  const possRoles = data[4714];

  const pushDataToClipboard = async () => {
    await window.navigator.clipboard.writeText(JSON.stringify(data));
  };

  return (
    <Grid container xs={12} direction="row">
      <Grid item xs={12}>
        <Typography size="18px" style={{ fontWeight: 800 }}>
          {uid}
        </Typography>
        <IndividualName uid={classFact.fact_uid} name={name} />
        <Synonyms synonymFacts={synFacts} />
        <Collection uid={definition[0].fact_uid} collection={collection} />
        <Typography size="12px" style={{ fontWeight: 500 }}>
          {classification && classification.length > 0
            ? `Classification:`
            : "error, concept seemingly not classified"}
        </Typography>
        {classification && classification.length > 0 ? (
          <Classification uids={classification} individualUID={uid} />
        ) : (
          "error, concept seemingly not classified"
        )}
        {value && <Value value={value} />}
        {definition && <Definition definitions={definition} />}
      </Grid>
      <Grid
        item
        xs={12}
        border={1}
        borderRadius={2}
        borderColor={"orange"}
        margin={1}
      >
        <IconButton onClick={pushDataToClipboard}>
          <CopyAllIcon />
        </IconButton>
      </Grid>

      {facts && facts.length > 0 && (
        <Box>
          <Typography size="12px" style={{ fontWeight: 600 }}>
            AllRelatedFacts:
          </Typography>
          {/*<TableContainer component={Paper}>
              <Table>
                <TableBody>{factTableRows}</TableBody>
              </Table>
            </TableContainer>*/}
          <WorkflowFactsVisualizer facts={facts} sparse={true} />
        </Box>
      )}
    </Grid>
  );
};

export default IndividualDetails;
