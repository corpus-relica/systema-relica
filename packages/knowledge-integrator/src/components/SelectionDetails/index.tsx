import React, { useState, useEffect } from "react";
import {
  Box,
  Button,
  Grommet,
  Table,
  TableRow,
  TableBody,
  TableCell,
  Text,
} from "grommet";
import ErrorBoundary from "../ErrorBoundary";
import { useStores } from "../../context/RootStoreContext";
import { observer } from "mobx-react";
import { getEntityType } from "../../RLCBaseClient";
import { useQuery } from "@tanstack/react-query";
import KindDetails from "./KindDetails";
import IndividualDetails from "./IndividualDetails";
import IntegerDetails from "./IntegerDetails";

const START_INT_UID_RANGE = 5000000000;

const SelectionDetails: React.FC = observer(() => {
  const { graphViewStore } = useStores();
  const { selectedNode } = graphViewStore;

  const { isLoading, error, data } = useQuery({
    queryKey: ["entityType", selectedNode],
    queryFn: () =>
      selectedNode
        ? getEntityType(selectedNode).then((res) => res)
        : Promise.resolve(null),
    enabled: !!selectedNode && selectedNode < START_INT_UID_RANGE, // This disables the query if selectedNode is null
  });

  if (!selectedNode) return <div>No entity selected</div>;
  if (isLoading) return <div>Loading...</div>;
  // if (isLoading || !selectedNode) return <div>Loading...</div>;
  if (error) return <div>Error: {error.message}</div>;

  let comp;
  if (selectedNode >= START_INT_UID_RANGE) {
    comp = <IntegerDetails />;
  } else if (data && (data === "kind" || data === "qualification")) {
    comp = <KindDetails />;
  } else if (data && data === "individual") {
    comp = <IndividualDetails />;
  } else {
    comp = <div>Unknown entity type</div>;
    console.log("Unknown entity type");
    console.log(data);
  }

  return (
    <ErrorBoundary>
      <Box>{comp}</Box>
    </ErrorBoundary>
  );
});

export default SelectionDetails;
