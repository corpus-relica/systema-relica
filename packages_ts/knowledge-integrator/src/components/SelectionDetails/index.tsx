import React, { useState, useEffect } from "react";
import { useStore } from "react-admin";
import { useQuery } from "@tanstack/react-query";
import { getEntityType } from "../../RLCBaseClient";
import KindDetails from "./KindDetails";
import IndividualDetails from "./IndividualDetails";

const START_INT_UID_RANGE = 5000000000;

const SelectionDetails = () => {
  const [selectedNode] = useStore("selectedNode");
  const [selectedEdge] = useStore("selectedEdge");

  const { isLoading, error, data } = useQuery({
    queryKey: ["entityType", selectedNode],
    queryFn: () =>
      selectedNode
        ? getEntityType(selectedNode).then((res) => res)
        : Promise.resolve(null),
    enabled: !!selectedNode && selectedNode < START_INT_UID_RANGE, // This disables the query if selectedNode is null
  });

  if (!selectedNode) return <div>No entity selected</div>;
  if (isLoading) return <div>Loading...</div>;
  // if (isLoading || !selectedNode) return <div>Loading...</div>;
  if (error) return <div>Error: {error.message}</div>;

  let comp;
  if (selectedNode >= START_INT_UID_RANGE) {
    comp = <div>Integer:{selectedNode}</div>; //<IntegerDetails />;
  } else if (data && (data === "kind" || data === "qualification")) {
    comp = <KindDetails />;
  } else if (data && data === "individual") {
    comp = <IndividualDetails />;
  } else {
    comp = <div>Unknown entity type</div>;
    console.log("Unknown entity type");
    console.log(data);
  }

  return comp;
};

export default SelectionDetails;
