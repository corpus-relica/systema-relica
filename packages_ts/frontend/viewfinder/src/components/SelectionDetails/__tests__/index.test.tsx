import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import SelectionDetails from '../index';
import { MockWebSocket } from '../../../tests/setup/websocketMock';
import { setupMockWebSocket, cleanupMockWebSocket } from '../../../tests/helpers/cacheTestHelpers';

// Mock react-admin
jest.mock('react-admin', () => ({
  useStore: jest.fn()
}));

// Mock the PortalClient
jest.mock('../../../io/PortalClient.js', () => ({
  portalClient: {
    getEntityType: jest.fn()
  }
}));

// Mock child components
jest.mock('../KindDetails/index.js', () => {
  return function MockKindDetails() {
    return <div data-testid="kind-details">Kind Details Component</div>;
  };
});

jest.mock('../IndividualDetails/index.js', () => {
  return function MockIndividualDetails() {
    return <div data-testid="individual-details">Individual Details Component</div>;
  };
});

describe('SelectionDetails', () => {
  let mockWs: MockWebSocket;
  let mockUseStore: jest.Mock;
  let mockPortalClient: any;
  let queryClient: QueryClient;

  const renderWithQueryClient = (component: React.ReactElement) => {
    return render(
      <QueryClientProvider client={queryClient}>
        {component}
      </QueryClientProvider>
    );
  };

  beforeEach(() => {
    mockWs = setupMockWebSocket();
    queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
        },
      },
    });

    mockUseStore = jest.fn();
    mockPortalClient = {
      getEntityType: jest.fn()
    };

    const { useStore } = require('react-admin');
    useStore.mockImplementation(mockUseStore);

    const { portalClient } = require('../../../io/PortalClient.js');
    Object.assign(portalClient, mockPortalClient);
  });

  afterEach(() => {
    cleanupMockWebSocket();
    jest.clearAllMocks();
    queryClient.clear();
  });

  describe('No Selection State', () => {
    it('displays no entity selected message when no node is selected', () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [null];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      renderWithQueryClient(<SelectionDetails />);

      expect(screen.getByText('No entity selected')).toBeInTheDocument();
    });
  });

  describe('Loading State', () => {
    it('displays loading message while fetching entity type', () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [123];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockImplementation(() => 
        new Promise(resolve => setTimeout(() => resolve({ type: 'kind' }), 100))
      );

      renderWithQueryClient(<SelectionDetails />);

      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });
  });

  describe('Error State', () => {
    it('displays error message when API call fails', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [123];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      const errorMessage = 'Failed to fetch entity type';
      mockPortalClient.getEntityType.mockRejectedValue(new Error(errorMessage));

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByText(`Error: ${errorMessage}`)).toBeInTheDocument();
      });
    });
  });

  describe('Integer Entity Handling', () => {
    it('displays integer entity for UIDs >= 5000000000', async () => {
      const integerUid = 5000000001;
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [integerUid];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByText(`Integer:${integerUid}`)).toBeInTheDocument();
      });

      // Should not call API for integer entities
      expect(mockPortalClient.getEntityType).not.toHaveBeenCalled();
    });

    it('handles edge case at integer boundary', async () => {
      const boundaryUid = 5000000000;
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [boundaryUid];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByText(`Integer:${boundaryUid}`)).toBeInTheDocument();
      });
    });
  });

  describe('Kind Entity Handling', () => {
    it('renders KindDetails for kind entity type', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [123];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockResolvedValue({ type: 'kind' });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByTestId('kind-details')).toBeInTheDocument();
        expect(screen.getByText('Kind Details Component')).toBeInTheDocument();
      });

      expect(mockPortalClient.getEntityType).toHaveBeenCalledWith(123);
    });

    it('renders KindDetails for qualification entity type', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [456];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockResolvedValue({ type: 'qualification' });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByTestId('kind-details')).toBeInTheDocument();
        expect(screen.getByText('Kind Details Component')).toBeInTheDocument();
      });

      expect(mockPortalClient.getEntityType).toHaveBeenCalledWith(456);
    });
  });

  describe('Individual Entity Handling', () => {
    it('renders IndividualDetails for individual entity type', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [789];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockResolvedValue({ type: 'individual' });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByTestId('individual-details')).toBeInTheDocument();
        expect(screen.getByText('Individual Details Component')).toBeInTheDocument();
      });

      expect(mockPortalClient.getEntityType).toHaveBeenCalledWith(789);
    });
  });

  describe('Unknown Entity Type Handling', () => {
    it('displays unknown entity type message for unrecognized types', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [999];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      const consoleSpy = jest.spyOn(console, 'log').mockImplementation();
      mockPortalClient.getEntityType.mockResolvedValue({ type: 'unknown' });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByText('Unknown entity type')).toBeInTheDocument();
      });

      expect(consoleSpy).toHaveBeenCalledWith('Unknown entity type');
      expect(consoleSpy).toHaveBeenCalledWith({ type: 'unknown' });

      consoleSpy.mockRestore();
    });

    it('handles null entity type response', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [999];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      const consoleSpy = jest.spyOn(console, 'log').mockImplementation();
      mockPortalClient.getEntityType.mockResolvedValue({ type: null });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(screen.getByText('Unknown entity type')).toBeInTheDocument();
      });

      expect(consoleSpy).toHaveBeenCalledWith('Unknown entity type');
      expect(consoleSpy).toHaveBeenCalledWith({ type: null });

      consoleSpy.mockRestore();
    });
  });

  describe('Query Behavior', () => {
    it('does not fetch entity type for null selectedNode', () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [null];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      renderWithQueryClient(<SelectionDetails />);

      expect(mockPortalClient.getEntityType).not.toHaveBeenCalled();
    });

    it('does not fetch entity type for integer UIDs', () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [5000000001];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      renderWithQueryClient(<SelectionDetails />);

      expect(mockPortalClient.getEntityType).not.toHaveBeenCalled();
    });

    it('fetches entity type for valid non-integer UIDs', async () => {
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [123];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockResolvedValue({ type: 'kind' });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(mockPortalClient.getEntityType).toHaveBeenCalledWith(123);
      });
    });
  });

  describe('Store Integration', () => {
    it('correctly reads selectedNode from store', async () => {
      const selectedNodeValue = 456;
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [selectedNodeValue];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockResolvedValue({ type: 'individual' });

      renderWithQueryClient(<SelectionDetails />);

      await waitFor(() => {
        expect(mockPortalClient.getEntityType).toHaveBeenCalledWith(selectedNodeValue);
      });
    });

    it('correctly reads selectedEdge from store', () => {
      const selectedEdgeValue = 789;
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [null];
        if (key === 'selectedEdge') return [selectedEdgeValue];
        return [null];
      });

      renderWithQueryClient(<SelectionDetails />);

      // Should still show no entity selected since selectedNode is null
      expect(screen.getByText('No entity selected')).toBeInTheDocument();
    });
  });

  describe('Component Lifecycle', () => {
    it('updates when selectedNode changes', async () => {
      const { rerender } = renderWithQueryClient(<SelectionDetails />);

      // Initially no selection
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [null];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      rerender(
        <QueryClientProvider client={queryClient}>
          <SelectionDetails />
        </QueryClientProvider>
      );

      expect(screen.getByText('No entity selected')).toBeInTheDocument();

      // Change to have a selection
      mockUseStore.mockImplementation((key: string) => {
        if (key === 'selectedNode') return [123];
        if (key === 'selectedEdge') return [null];
        return [null];
      });

      mockPortalClient.getEntityType.mockResolvedValue({ type: 'kind' });

      rerender(
        <QueryClientProvider client={queryClient}>
          <SelectionDetails />
        </QueryClientProvider>
      );

      await waitFor(() => {
        expect(screen.getByTestId('kind-details')).toBeInTheDocument();
      });
    });
  });

  it('takes a snapshot with no selection', () => {
    mockUseStore.mockImplementation((key: string) => {
      if (key === 'selectedNode') return [null];
      if (key === 'selectedEdge') return [null];
      return [null];
    });

    const { container } = renderWithQueryClient(<SelectionDetails />);
    expect(container).toMatchSnapshot();
  });

  it('takes a snapshot with integer entity', () => {
    mockUseStore.mockImplementation((key: string) => {
      if (key === 'selectedNode') return [5000000001];
      if (key === 'selectedEdge') return [null];
      return [null];
    });

    const { container } = renderWithQueryClient(<SelectionDetails />);
    expect(container).toMatchSnapshot();
  });
});