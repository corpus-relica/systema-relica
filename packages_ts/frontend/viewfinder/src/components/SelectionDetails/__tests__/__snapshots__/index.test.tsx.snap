// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionDetails takes a snapshot with integer entity 1`] = `
<div>
  <div>
    Integer:
    5000000001
  </div>
</div>
`;

exports[`SelectionDetails takes a snapshot with no selection 1`] = `
<div>
  <div>
    No entity selected
  </div>
</div>
`;
